pkg/apply/solver/solver.go

```diff
@@ -86,74 +87,139 @@ func (tq *TaskQueue) ToActionGroups() []event.ActionGroup {
-	applyObjs = t.Collector.FilterInvalidObjects(applyObjs)
-	klog.V(2).Infoln("adding inventory add task (%d objects)", len(applyObjs))
-	t.tasks = append(t.tasks, &task.InvAddTask{
-		TaskName:  fmt.Sprintf("inventory-add-%d", t.invAddCounter),
-		InvClient: t.InvClient,
-		InvInfo:   inv,
-		Objects:   applyObjs,
-		DryRun:    dryRun,
-	})
-	t.invAddCounter++
+// WithApplyObjects sets the apply objects and returns the builder for chaining.
+func (t *TaskQueueBuilder) WithApplyObjects(applyObjs object.UnstructuredSet) *TaskQueueBuilder {
+	t.applyObjs = applyObjs
 	return t
 }
 
-// AppendInvSetTask appends an inventory set task to the task queue.
-// Returns a pointer to the Builder to chain function calls.
-func (t *TaskQueueBuilder) AppendInvSetTask(inv inventory.Info, dryRun common.DryRunStrategy) *TaskQueueBuilder {
-	klog.V(2).Infoln("adding inventory set task")
-	prevInvIds, _ := t.InvClient.GetClusterObjs(inv)
-	t.tasks = append(t.tasks, &task.InvSetTask{
-		TaskName:      fmt.Sprintf("inventory-set-%d", t.invSetCounter),
-		InvClient:     t.InvClient,
-		InvInfo:       inv,
-		PrevInventory: prevInvIds,
-		DryRun:        dryRun,
-	})
-	t.invSetCounter++
+// WithPruneObjects sets the prune objects and returns the builder for chaining.
+func (t *TaskQueueBuilder) WithPruneObjects(pruneObjs object.UnstructuredSet) *TaskQueueBuilder {
+	t.pruneObjs = pruneObjs
 	return t
 }
 
-// AppendDeleteInvTask appends to the task queue a task to delete the inventory object.
-// Returns a pointer to the Builder to chain function calls.
-func (t *TaskQueueBuilder) AppendDeleteInvTask(inv inventory.Info, dryRun common.DryRunStrategy) *TaskQueueBuilder {
-	klog.V(2).Infoln("adding delete inventory task")
-	t.tasks = append(t.tasks, &task.DeleteInvTask{
-		TaskName:  fmt.Sprintf("delete-inventory-%d", t.deleteInvCounter),
-		InvClient: t.InvClient,
-		InvInfo:   inv,
-		DryRun:    dryRun,
-	})
-	t.deleteInvCounter++
-	return t
+// Build returns the queue of tasks that have been created
+func (t *TaskQueueBuilder) Build(o Options) *TaskQueue {
+	var tasks []taskrunner.Task
+
+	// reset counters
+	t.applyCounter = 0
+	t.pruneCounter = 0
+	t.waitCounter = 0
+
+	applyObjs := t.Collector.FilterInvalidObjects(t.applyObjs)
+	pruneObjs := t.Collector.FilterInvalidObjects(t.pruneObjs)
+
+	if len(applyObjs) > 0 {
+		klog.V(2).Infoln("adding inventory add task (%d objects)", len(applyObjs))
+		tasks = append(tasks, &task.InvAddTask{
+			TaskName:  "inventory-add-0",
+			InvClient: t.InvClient,
+			InvInfo:   t.invInfo,
+			Objects:   applyObjs,
+			DryRun:    o.DryRunStrategy,
+		})
+	}
+
+	// Use the "depends-on" annotation to create a graph, ands sort the
+	// objects to apply into sets using a topological sort.
+	if len(applyObjs) > 0 {
+		applySets, err := graph.SortObjs(applyObjs)
+		if err != nil {
+			t.Collector.Collect(err)
+		}
+		for _, applySet := range applySets {
+			applySet = t.Collector.FilterInvalidObjects(applySet)
```

Comment:
One thing that I don't quite understand is why filtering the applySet in addition to the filtering in line 130-131.

Reply:
you're right, it's redundant. I'll remove these ones.

Reply:
turns out this is required, because graphing and sorting can produce additional invalid objects.

Reply:
I added comments to explain

Reply:
Ack